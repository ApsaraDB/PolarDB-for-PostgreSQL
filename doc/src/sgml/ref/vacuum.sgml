<!--
doc/src/sgml/ref/vacuum.sgml
PostgreSQL documentation
-->

<refentry id="sql-vacuum">
 <indexterm zone="sql-vacuum">
  <primary>VACUUM</primary>
 </indexterm>

 <refmeta>
  <refentrytitle>VACUUM</refentrytitle>
  <manvolnum>7</manvolnum>
  <refmiscinfo>SQL - Language Statements</refmiscinfo>
 </refmeta>

 <refnamediv>
  <refname>VACUUM</refname>
  <refpurpose>garbage-collect and optionally analyze a database</refpurpose>
 </refnamediv>

 <refsynopsisdiv>
<synopsis>
VACUUM [ ( <replaceable class="parameter">option</replaceable> [, ...] ) ] [ <replaceable class="parameter">table_and_columns</replaceable> [, ...] ]
VACUUM [ FULL ] [ FREEZE ] [ VERBOSE ] [ ANALYZE ] [ <replaceable class="parameter">table_and_columns</replaceable> [, ...] ]

<phrase>where <replaceable class="parameter">option</replaceable> can be one of:</phrase>

    FULL
    FREEZE
    VERBOSE
    ANALYZE
    DISABLE_PAGE_SKIPPING

<phrase>and <replaceable class="parameter">table_and_columns</replaceable> is:</phrase>

    <replaceable class="parameter">table_name</replaceable> [ ( <replaceable class="parameter">column_name</replaceable> [, ...] ) ]
</synopsis>
 </refsynopsisdiv>

 <refsect1>
  <title>Description</title>

  <para>
   <command>VACUUM</command> reclaims storage occupied by dead tuples.
   In normal <productname>PostgreSQL</productname> operation, tuples that
   are deleted or obsoleted by an update are not physically removed from
   their table; they remain present until a <command>VACUUM</command> is
   done.  Therefore it's necessary to do <command>VACUUM</command>
   periodically, especially on frequently-updated tables.
  </para>

  <para>
   Without a <replaceable class="parameter">table_and_columns</replaceable>
   list, <command>VACUUM</command> processes every table and materialized view
   in the current database that the current user has permission to vacuum.
   With a list, <command>VACUUM</command> processes only those table(s).
  </para>

  <para>
   <command>VACUUM ANALYZE</command> performs a <command>VACUUM</command>
   and then an <command>ANALYZE</command> for each selected table.  This
   is a handy combination form for routine maintenance scripts.  See
   <xref linkend="sql-analyze"/>
   for more details about its processing.
  </para>

  <para>
   Plain <command>VACUUM</command> (without <literal>FULL</literal>) simply reclaims
   space and makes it
   available for re-use.  This form of the command can operate in parallel
   with normal reading and writing of the table, as an exclusive lock
   is not obtained.  However, extra space is not returned to the operating
   system (in most cases); it's just kept available for re-use within the
   same table.  <command>VACUUM FULL</command> rewrites the entire contents
   of the table into a new disk file with no extra space, allowing unused
   space to be returned to the operating system.  This form is much slower and
   requires an exclusive lock on each table while it is being processed.
  </para>

  <para>
   When the option list is surrounded by parentheses, the options can be
   written in any order.  Without parentheses, options must be specified
   in exactly the order shown above.
   The parenthesized syntax was added in
   <productname>PostgreSQL</productname> 9.0;  the unparenthesized
   syntax is deprecated.
  </para>
 </refsect1>

 <refsect1>
  <title>Parameters</title>

  <variablelist>
   <varlistentry>
    <term><literal>FULL</literal></term>
    <listitem>
     <para>
      Selects <quote>full</quote> vacuum, which can reclaim more
      space, but takes much longer and exclusively locks the table.
      This method also requires extra disk space, since it writes a
      new copy of the table and doesn't release the old copy until
      the operation is complete.  Usually this should only be used when a
      significant amount of space needs to be reclaimed from within the table.
     </para>
    </listitem>
   </varlistentry>

   <varlistentry>
    <term><literal>FREEZE</literal></term>
    <listitem>
     <para>
      Selects aggressive <quote>freezing</quote> of tuples.
      Specifying <literal>FREEZE</literal> is equivalent to performing
      <command>VACUUM</command> with the
      <xref linkend="guc-vacuum-freeze-min-age"/> and
      <xref linkend="guc-vacuum-freeze-table-age"/> parameters
      set to zero.  Aggressive freezing is always performed when the
      table is rewritten, so this option is redundant when <literal>FULL</literal>
      is specified.
     </para>
    </listitem>
   </varlistentry>

   <varlistentry>
    <term><literal>VERBOSE</literal></term>
    <listitem>
     <para>
      Prints a detailed vacuum activity report for each table.
     </para>
    </listitem>
   </varlistentry>

   <varlistentry>
    <term><literal>ANALYZE</literal></term>
    <listitem>
     <para>
      Updates statistics used by the planner to determine the most
      efficient way to execute a query.
     </para>
    </listitem>
   </varlistentry>

   <varlistentry>
    <term><literal>DISABLE_PAGE_SKIPPING</literal></term>
    <listitem>
     <para>
      Normally, <command>VACUUM</command> will skip pages based on the <link
      linkend="vacuum-for-visibility-map">visibility map</link>.  Pages where
      all tuples are known to be frozen can always be skipped, and those
      where all tuples are known to be visible to all transactions may be
      skipped except when performing an aggressive vacuum.  Furthermore,
      except when performing an aggressive vacuum, some pages may be skipped
      in order to avoid waiting for other sessions to finish using them.
      This option disables all page-skipping behavior, and is intended to
      be used only when the contents of the visibility map are
      suspect, which should happen only if there is a hardware or software
      issue causing database corruption.
     </para>
    </listitem>
   </varlistentry>

   <varlistentry>
    <term><replaceable class="parameter">table_name</replaceable></term>
    <listitem>
     <para>
      The name (optionally schema-qualified) of a specific table or
      materialized view to vacuum.  If the specified table is a partitioned
      table, all of its leaf partitions are vacuumed.
     </para>
    </listitem>
   </varlistentry>

   <varlistentry>
    <term><replaceable class="parameter">column_name</replaceable></term>
    <listitem>
     <para>
      The name of a specific column to analyze. Defaults to all columns.
      If a column list is specified, <literal>ANALYZE</literal> must also be
      specified.
     </para>
    </listitem>
   </varlistentry>
  </variablelist>
 </refsect1>

 <refsect1>
  <title>Outputs</title>

   <para>
    When <literal>VERBOSE</literal> is specified, <command>VACUUM</command> emits
    progress messages to indicate which table is currently being
    processed.  Various statistics about the tables are printed as well.
   </para>
 </refsect1>

 <refsect1>
  <title>Notes</title>

   <para>
    To vacuum a table, one must ordinarily be the table's owner or a
    superuser.  However, database owners are allowed to
    vacuum all tables in their databases, except shared catalogs.
    (The restriction for shared catalogs means that a true database-wide
    <command>VACUUM</command> can only be performed by a superuser.)
    <command>VACUUM</command> will skip over any tables that the calling user
    does not have permission to vacuum.
   </para>

   <para>
    <command>VACUUM</command> cannot be executed inside a transaction block.
   </para>

   <para>
    For tables with <acronym>GIN</acronym> indexes, <command>VACUUM</command> (in
    any form) also completes any pending index insertions, by moving pending
    index entries to the appropriate places in the main <acronym>GIN</acronym> index
    structure.  See <xref linkend="gin-fast-update"/> for details.
   </para>

   <para>
    We recommend that active production databases be
    vacuumed frequently (at least nightly), in order to
    remove dead rows. After adding or deleting a large number
    of rows, it might be a good idea to issue a <command>VACUUM
    ANALYZE</command> command for the affected table. This will update the
    system catalogs with
    the results of all recent changes, and allow the
    <productname>PostgreSQL</productname> query planner to make better
    choices in planning queries.
   </para>

   <para>
    The <option>FULL</option> option is not recommended for routine use,
    but might be useful in special cases.  An example is when you have deleted
    or updated most of the rows in a table and would like the table to
    physically shrink to occupy less disk space and allow faster table
    scans. <command>VACUUM FULL</command> will usually shrink the table
    more than a plain <command>VACUUM</command> would.
   </para>

   <para>
    <command>VACUUM</command> causes a substantial increase in I/O traffic,
    which might cause poor performance for other active sessions.  Therefore,
    it is sometimes advisable to use the cost-based vacuum delay feature.
    See <xref linkend="runtime-config-resource-vacuum-cost"/> for details.
   </para>

   <para>
    <productname>PostgreSQL</productname> includes an <quote>autovacuum</quote>
    facility which can automate routine vacuum maintenance.  For more
    information about automatic and manual vacuuming, see
    <xref linkend="routine-vacuuming"/>.
   </para>
 </refsect1>

 <refsect1>
  <title>Examples</title>

  <para>
   To clean a single table <literal>onek</literal>, analyze it for
   the optimizer and print a detailed vacuum activity report:

<programlisting>
VACUUM (VERBOSE, ANALYZE) onek;
</programlisting></para>
 </refsect1>

 <refsect1>
  <title>Compatibility</title>

  <para>
   There is no <command>VACUUM</command> statement in the SQL standard.
  </para>
 </refsect1>

 <refsect1>
  <title>See Also</title>

  <simplelist type="inline">
   <member><xref linkend="app-vacuumdb"/></member>
   <member><xref linkend="runtime-config-resource-vacuum-cost"/></member>
   <member><xref linkend="autovacuum"/></member>
  </simplelist>
 </refsect1>
</refentry>
