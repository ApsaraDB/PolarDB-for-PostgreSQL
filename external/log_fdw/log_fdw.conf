logging_collector = on
# log_destination = 'stderr, csvlog'
log_destination = 'stderr'
# these ensure stability of test results:
log_rotation_age = 0
lc_messages = 'C'
