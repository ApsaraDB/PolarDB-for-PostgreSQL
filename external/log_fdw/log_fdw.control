# log_fdw extension
comment = 'foreign-data wrapper for Postgres log file access'
default_version = '1.4'
module_pathname = '$libdir/log_fdw'
relocatable = true
