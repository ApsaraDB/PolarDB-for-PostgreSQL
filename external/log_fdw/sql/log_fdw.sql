--
-- Test foreign-data wrapper log_fdw.
--

-- Install log_fdw
CREATE EXTENSION log_fdw;

-- Create foreign server
CREATE SERVER log_fdw_server FOREIGN DATA WRAPPER log_fdw;

-- Invalid option name
CREATE FOREIGN TABLE log_fdw_ftbl_err () SERVER log_fdw_server OPTIONS (foo 'bar');  -- ERROR

-- Invalid option value i.e. absolute path as value
CREATE FOREIGN TABLE log_fdw_ftbl_err () SERVER log_fdw_server OPTIONS (filename '/foo/bar');  -- ERROR

-- Option provided more than once
CREATE FOREIGN TABLE log_fdw_ftbl_err () SERVER log_fdw_server OPTIONS (filename 'foo', filename 'bar');  -- ERROR

-- Provided no option
CREATE FOREIGN TABLE log_fdw_ftbl_err () SERVER log_fdw_server;  -- ERROR

-- Check permissions
CREATE ROLE regress_log_fdw_nsuperuser;

-- A non-superuser has no permission to use log_fdw by default
SELECT has_function_privilege('regress_log_fdw_nsuperuser',
    'create_foreign_table_for_log_file(text, text, text)', 'EXECUTE'); -- no

SELECT has_function_privilege('regress_log_fdw_nsuperuser',
    'list_postgres_log_files()', 'EXECUTE'); -- no

-- A non-superuser can be granted permission to use log_fdw
GRANT EXECUTE ON FUNCTION create_foreign_table_for_log_file(text, text, text)
    TO regress_log_fdw_nsuperuser;

GRANT EXECUTE ON FUNCTION list_postgres_log_files()
    TO regress_log_fdw_nsuperuser;

SELECT has_function_privilege('regress_log_fdw_nsuperuser',
    'create_foreign_table_for_log_file(text, text, text)', 'EXECUTE'); -- yes

SELECT has_function_privilege('regress_log_fdw_nsuperuser',
    'list_postgres_log_files()', 'EXECUTE'); -- yes

-- Ensure that something gets logged to the server log file
SELECT 1/0;

-- Ensure there exists at least one server log file to work on
SElECT pg_rotate_logfile();
DO $$
DECLARE
    log_file INT := 0;
    loops INT := 0;
BEGIN
    LOOP
        log_file := COUNT(*) FROM list_postgres_log_files();
        -- -- Check if both .csv and .log files get created
        -- Check if .log file get created
        IF log_file >= 1 OR loops > 180 * 100 THEN EXIT; END IF;
        PERFORM pg_sleep(0.01);
        loops := loops + 1;
    END LOOP;
END
$$;

-- Check if .csv file gets created
-- SELECT COUNT(*) > 0 AS OK FROM list_postgres_log_files()
--     WHERE file_name LIKE '%.csv';

-- Check if .log file gets created
SELECT COUNT(*) > 0 AS OK FROM list_postgres_log_files()
    WHERE file_name LIKE '%.log';

-- Get .csv server log file name and create a foreign table for it
-- DO $$
-- DECLARE
--     log_file_name TEXT;
-- BEGIN
--     log_file_name := file_name FROM list_postgres_log_files()
--              WHERE file_name LIKE '%.csv' ORDER BY file_name ASC LIMIT 1;
--     PERFORM create_foreign_table_for_log_file('log_fdw_ftbl_csv','log_fdw_server', log_file_name);
-- END
-- $$;

-- Ensure that the foreign table gets created
-- SELECT COUNT(*) > 0 AS OK FROM pg_foreign_table WHERE ftrelid = 'log_fdw_ftbl_csv'::regclass;

-- Ensure that the foreign table has read data from .csv server log file
-- SELECT COUNT(*) > 0 AS OK FROM log_fdw_ftbl_csv;

-- Get .log server log file name and create a foreign table for it
DO $$
DECLARE
    log_file_name TEXT;
BEGIN
    log_file_name := file_name FROM list_postgres_log_files()
             WHERE file_name LIKE '%.log' ORDER BY file_name ASC LIMIT 1;
    PERFORM create_foreign_table_for_log_file('log_fdw_ftbl_log','log_fdw_server', log_file_name);
END
$$;

-- Ensure that the foreign table gets created
SELECT COUNT(*) > 0 AS OK FROM pg_foreign_table WHERE ftrelid = 'log_fdw_ftbl_log'::regclass;

-- Ensure that the foreign table has read data from .log server log file
SELECT COUNT(*) > 0 AS OK FROM log_fdw_ftbl_log;

-- Clean up
DROP EXTENSION log_fdw CASCADE;
