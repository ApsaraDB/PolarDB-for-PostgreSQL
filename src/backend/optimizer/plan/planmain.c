/*-------------------------------------------------------------------------
 *
 * planmain.c
 *	  Routines to plan a single query
 *
 * What's in a name, anyway?  The top-level entry point of the planner/
 * optimizer is over in planner.c, not here as you might think from the
 * file name.  But this is the main code for planning a basic join operation,
 * shorn of features like subselects, inheritance, aggregates, grouping,
 * and so on.  (Those are the things planner.c deals with.)
 *
 * Portions Copyright (c) 1996-2018, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 *
 * IDENTIFICATION
 *	  src/backend/optimizer/plan/planmain.c
 *
 *-------------------------------------------------------------------------
 */
#include "postgres.h"

#include "optimizer/clauses.h"
#include "optimizer/orclauses.h"
#include "optimizer/pathnode.h"
#include "optimizer/paths.h"
#include "optimizer/placeholder.h"
#include "optimizer/planmain.h"
#include "optimizer/cost.h"
#include "px/px_vars.h"

/* POLAR px */
#include "utils/guc.h"

/*
 * query_planner
 *	  Generate a path (that is, a simplified plan) for a basic query,
 *	  which may involve joins but not any fancier features.
 *
 * Since query_planner does not handle the toplevel processing (grouping,
 * sorting, etc) it cannot select the best path by itself.  Instead, it
 * returns the RelOptInfo for the top level of joining, and the caller
 * (grouping_planner) can choose among the surviving paths for the rel.
 *
 * root describes the query to plan
 * tlist is the target list the query should produce
 *		(this is NOT necessarily root->parse->targetList!)
 * qp_callback is a function to compute query_pathkeys once it's safe to do so
 * qp_extra is optional extra data to pass to qp_callback
 *
 * Note: the PlannerInfo node also includes a query_pathkeys field, which
 * tells query_planner the sort order that is desired in the final output
 * plan.  This value is *not* available at call time, but is computed by
 * qp_callback once we have completed merging the query's equivalence classes.
 * (We cannot construct canonical pathkeys until that's done.)
 */
RelOptInfo *
query_planner(PlannerInfo *root, List *tlist,
			  query_pathkeys_callback qp_callback, void *qp_extra)
{
	Query	   *parse = root->parse;
	List	   *joinlist;
	RelOptInfo *final_rel;
	Index		rti;
	double		total_pages;

	/*
	 * If the query has an empty join tree, then it's something easy like
	 * "SELECT 2+2;" or "INSERT ... VALUES()".  Fall through quickly.
	 */
	if (parse->jointree->fromlist == NIL)
	{
		/* We need a dummy joinrel to describe the empty set of baserels */
		final_rel = build_empty_join_rel(root);

		/*
		 * If query allows parallelism in general, check whether the quals are
		 * parallel-restricted.  (We need not check final_rel->reltarget
		 * because it's empty at this point.  Anything parallel-restricted in
		 * the query tlist will be dealt with later.)
		 */
		if (root->glob->parallelModeOK)
			final_rel->consider_parallel =
				is_parallel_safe(root, parse->jointree->quals);

		/* The only path for it is a trivial Result path */
		add_path(final_rel, (Path *)
				 create_result_path(root, final_rel,
									final_rel->reltarget,
									(List *) parse->jointree->quals));

		/* Select cheapest path (pretty easy in this case...) */
		set_cheapest(final_rel);

		/*
		 * We still are required to call qp_callback, in case it's something
		 * like "SELECT 2+2 ORDER BY 1".
		 */
		root->canon_pathkeys = NIL;
		(*qp_callback) (root, qp_extra);

		return final_rel;
	}

	/*
	 * Init planner lists to empty.
	 *
	 * NOTE: append_rel_list was set up by subquery_planner, so do not touch
	 * here.
	 */
	root->join_rel_list = NIL;
	root->join_rel_hash = NULL;
	root->join_rel_level = NULL;
	root->join_cur_level = 0;
	root->canon_pathkeys = NIL;
	root->left_join_clauses = NIL;
	root->right_join_clauses = NIL;
	root->full_join_clauses = NIL;
	root->join_info_list = NIL;
	root->placeholder_list = NIL;
	root->fkey_list = NIL;
	root->initial_rels = NIL;

	/*
	 * Make a flattened version of the rangetable for faster access (this is
	 * OK because the rangetable won't change any more), and set up an empty
	 * array for indexing base relations.
	 */
	setup_simple_rel_arrays(root);

	/*
	 * Populate append_rel_array with each AppendRelInfo to allow direct
	 * lookups by child relid.
	 */
	setup_append_rel_array(root);

	/*
	 * Construct RelOptInfo nodes for all base relations in query, and
	 * indirectly for all appendrel member relations ("other rels").  This
	 * will give us a RelOptInfo for every "simple" (non-join) rel involved in
	 * the query.
	 *
	 * Note: the reason we find the rels by searching the jointree and
	 * appendrel list, rather than just scanning the rangetable, is that the
	 * rangetable may contain RTEs for rels not actively part of the query,
	 * for example views.  We don't want to make RelOptInfos for them.
	 */
	add_base_rels_to_query(root, (Node *) parse->jointree);

	/*
	 * Examine the targetlist and join tree, adding entries to baserel
	 * targetlists for all referenced Vars, and generating PlaceHolderInfo
	 * entries for all referenced PlaceHolderVars.  Restrict and join clauses
	 * are added to appropriate lists belonging to the mentioned relations. We
	 * also build EquivalenceClasses for provably equivalent expressions. The
	 * SpecialJoinInfo list is also built to hold information about join order
	 * restrictions.  Finally, we form a target joinlist for make_one_rel() to
	 * work from.
	 */
	build_base_rel_tlists(root, tlist);

	find_placeholders_in_jointree(root);

	find_lateral_references(root);

	joinlist = deconstruct_jointree(root);

	/*
	 * Reconsider any postponed outer-join quals now that we have built up
	 * equivalence classes.  (This could result in further additions or
	 * mergings of classes.)
	 */
	reconsider_outer_join_clauses(root);

	/*
	 * If we formed any equivalence classes, generate additional restriction
	 * clauses as appropriate.  (Implied join clauses are formed on-the-fly
	 * later.)
	 */
	generate_base_implied_equalities(root);

	/*
	 * We have completed merging equivalence sets, so it's now possible to
	 * generate pathkeys in canonical form; so compute query_pathkeys and
	 * other pathkeys fields in PlannerInfo.
	 */
	(*qp_callback) (root, qp_extra);

	/*
	 * Examine any "placeholder" expressions generated during subquery pullup.
	 * Make sure that the Vars they need are marked as needed at the relevant
	 * join level.  This must be done before join removal because it might
	 * cause Vars or placeholders to be needed above a join when they weren't
	 * so marked before.
	 */
	fix_placeholder_input_needed_levels(root);

	/*
	 * Remove any useless outer joins.  Ideally this would be done during
	 * jointree preprocessing, but the necessary information isn't available
	 * until we've built baserel data structures and classified qual clauses.
	 */
	joinlist = remove_useless_joins(root, joinlist);

	/*
	 * Also, reduce any semijoins with unique inner rels to plain inner joins.
	 * Likewise, this can't be done until now for lack of needed info.
	 */
	reduce_unique_semijoins(root);

	/*
	 * Now distribute "placeholders" to base rels as needed.  This has to be
	 * done after join removal because removal could change whether a
	 * placeholder is evaluable at a base rel.
	 */
	add_placeholders_to_base_rels(root);

	/*
	 * Construct the lateral reference sets now that we have finalized
	 * PlaceHolderVar eval levels.
	 */
	create_lateral_join_info(root);

	/*
	 * Match foreign keys to equivalence classes and join quals.  This must be
	 * done after finalizing equivalence classes, and it's useful to wait till
	 * after join removal so that we can skip processing foreign keys
	 * involving removed relations.
	 */
	match_foreign_keys_to_quals(root);

	/*
	 * Look for join OR clauses that we can extract single-relation
	 * restriction OR clauses from.
	 */
	extract_restriction_or_clauses(root);

	/*
	 * We should now have size estimates for every actual table involved in
	 * the query, and we also know which if any have been deleted from the
	 * query by join removal; so we can compute total_table_pages.
	 *
	 * Note that appendrels are not double-counted here, even though we don't
	 * bother to distinguish RelOptInfos for appendrel parents, because the
	 * parents will still have size zero.
	 *
	 * XXX if a table is self-joined, we will count it once per appearance,
	 * which perhaps is the wrong thing ... but that's not completely clear,
	 * and detecting self-joins here is difficult, so ignore it for now.
	 */
	total_pages = 0;
	for (rti = 1; rti < root->simple_rel_array_size; rti++)
	{
		RelOptInfo *brel = root->simple_rel_array[rti];

		if (brel == NULL)
			continue;

		Assert(brel->relid == rti); /* sanity check on array */

		if (IS_SIMPLE_REL(brel))
			total_pages += (double) brel->pages;
	}
	root->total_table_pages = total_pages;

	/*
	 * Ready to do the primary planning.
	 */
	final_rel = make_one_rel(root, joinlist);

	/* Check that we got at least one usable path */
	if (!final_rel || !final_rel->cheapest_total_path ||
		final_rel->cheapest_total_path->param_info != NULL)
		elog(ERROR, "failed to construct the join relation");

	return final_rel;
}

/* POLAR px*/
/*
 * Default configuration information
 */
PlannerConfig *DefaultPlannerConfig(void)
{
	PlannerConfig *c1 = (PlannerConfig *) palloc(sizeof(PlannerConfig));
	c1->enable_sort = enable_sort;
	c1->enable_hashagg = enable_hashagg;
	c1->enable_groupagg = enable_groupagg;
	c1->enable_nestloop = enable_nestloop;
	c1->enable_mergejoin = enable_mergejoin;
	c1->enable_hashjoin = enable_hashjoin;
	c1->px_enable_hashjoin_size_heuristic = px_enable_hashjoin_size_heuristic;
	c1->px_enable_predicate_propagation = px_enable_predicate_propagation;
	c1->constraint_exclusion = constraint_exclusion;

	c1->px_enable_minmax_optimization = px_enable_minmax_optimization;
	c1->px_enable_multiphase_agg = px_enable_multiphase_agg;
	c1->px_enable_preunique = px_enable_preunique;
	c1->px_eager_preunique = px_eager_preunique;
	c1->px_hashagg_streambottom = px_hashagg_streambottom;
	c1->px_enable_agg_distinct = px_enable_agg_distinct;
	c1->px_enable_dqa_pruning = px_enable_dqa_pruning;
	c1->px_eager_dqa_pruning = px_eager_dqa_pruning;
	c1->px_eager_one_phase_agg = px_eager_one_phase_agg;
	c1->px_eager_two_phase_agg = px_eager_two_phase_agg;
	c1->px_enable_sort_distinct = px_enable_sort_distinct;

	c1->px_enable_direct_dispatch = px_enable_direct_dispatch;

	c1->px_cte_sharing = px_cte_sharing;

	c1->honor_order_by = true;

	c1->is_under_subplan = false;

	return c1;
}
