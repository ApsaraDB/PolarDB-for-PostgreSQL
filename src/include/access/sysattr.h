/*-------------------------------------------------------------------------
 *
 * sysattr.h
 *	  POSTGRES system attribute definitions.
 *
 *
 * Portions Copyright (c) 1996-2018, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 * src/include/access/sysattr.h
 *
 *-------------------------------------------------------------------------
 */
#ifndef SYSATTR_H
#define SYSATTR_H

/* POLAR px */
#include "postgres.h"

#include "px/px_vars.h"

/*
 * Attribute numbers for the system-defined attributes
 */
#define SelfItemPointerAttributeNumber			(-1)
#define ObjectIdAttributeNumber					(-2)
#define MinTransactionIdAttributeNumber			(-3)
#define MinCommandIdAttributeNumber				(-4)
#define MaxTransactionIdAttributeNumber			(-5)
#define MaxCommandIdAttributeNumber				(-6)
#define TableOidAttributeNumber					(-7)
/* POLAR px */
#define PxWorkerIdAttributeNumber			    (-8)
#define RootSelfItemPointerAttributeNumber      (-9)
#define FirstLowInvalidHeapAttributeNumber      (-10)
/* POLAR end */
#endif							/* SYSATTR_H */
