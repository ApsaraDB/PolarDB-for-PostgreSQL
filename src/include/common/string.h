/*
 *	string.h
 *		string handling helpers
 *
 *	Portions Copyright (c) 1996-2025, PostgreSQL Global Development Group
 *	Portions Copyright (c) 1994, Regents of the University of California
 *
 *	src/include/common/string.h
 */
#ifndef COMMON_STRING_H
#define COMMON_STRING_H

#include <signal.h>

typedef struct StringInfoData *StringInfo;	/* avoid including stringinfo.h
											 * here */

typedef struct PromptInterruptContext
{
	/* To avoid including <setjmp.h> here, jmpbuf is declared "void *" */
	void	   *jmpbuf;			/* existing longjmp buffer */
	volatile sig_atomic_t *enabled; /* flag that enables longjmp-on-interrupt */
	bool		canceled;		/* indicates whether cancellation occurred */
} PromptInterruptContext;

/* functions in src/common/string.c */
extern bool pg_str_endswith(const char *str, const char *end);
extern int	strtoint(const char *restrict str, char **restrict endptr,
					 int base);
extern char *pg_clean_ascii(const char *str, int alloc_flags);
extern int	pg_strip_crlf(char *str);
extern bool pg_is_ascii(const char *str);

/* functions in src/common/pg_get_line.c */
extern char *pg_get_line(FILE *stream, PromptInterruptContext *prompt_ctx);
extern bool pg_get_line_buf(FILE *stream, StringInfo buf);
extern bool pg_get_line_append(FILE *stream, StringInfo buf,
							   PromptInterruptContext *prompt_ctx);

/* functions in src/common/sprompt.c */
extern char *simple_prompt(const char *prompt, bool echo);
extern char *simple_prompt_extended(const char *prompt, bool echo,
									PromptInterruptContext *prompt_ctx);

#endif							/* COMMON_STRING_H */
