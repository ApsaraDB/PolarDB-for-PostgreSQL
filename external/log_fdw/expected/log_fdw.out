--
-- Test foreign-data wrapper log_fdw.
--
-- Install log_fdw
CREATE EXTENSION log_fdw;
-- Create foreign server
CREATE SERVER log_fdw_server FOREIGN DATA WRAPPER log_fdw;
-- Invalid option name
CREATE FOREIGN TABLE log_fdw_ftbl_err () SERVER log_fdw_server OPTIONS (foo 'bar');  -- ERROR
ERROR:  invalid option "foo"
HINT:  Valid options in this context are: filename
-- Invalid option value i.e. absolute path as value
CREATE FOREIGN TABLE log_fdw_ftbl_err () SERVER log_fdw_server OPTIONS (filename '/foo/bar');  -- ERROR
ERROR:  absolute path is not allowed as filename for log_fdw foreign tables
-- Option provided more than once
CREATE FOREIGN TABLE log_fdw_ftbl_err () SERVER log_fdw_server OPTIONS (filename 'foo', filename 'bar');  -- ERROR
ERROR:  option "filename" provided more than once
-- Provided no option
CREATE FOREIGN TABLE log_fdw_ftbl_err () SERVER log_fdw_server;  -- ERROR
ERROR:  filename is required for log_fdw foreign tables
-- Check permissions
CREATE ROLE regress_log_fdw_nsuperuser;
-- A non-superuser has no permission to use log_fdw by default
SELECT has_function_privilege('regress_log_fdw_nsuperuser',
    'create_foreign_table_for_log_file(text, text, text)', 'EXECUTE'); -- no
 has_function_privilege 
------------------------
 f
(1 row)

SELECT has_function_privilege('regress_log_fdw_nsuperuser',
    'list_postgres_log_files()', 'EXECUTE'); -- no
 has_function_privilege 
------------------------
 f
(1 row)

-- A non-superuser can be granted permission to use log_fdw
GRANT EXECUTE ON FUNCTION create_foreign_table_for_log_file(text, text, text)
    TO regress_log_fdw_nsuperuser;
GRANT EXECUTE ON FUNCTION list_postgres_log_files()
    TO regress_log_fdw_nsuperuser;
SELECT has_function_privilege('regress_log_fdw_nsuperuser',
    'create_foreign_table_for_log_file(text, text, text)', 'EXECUTE'); -- yes
 has_function_privilege 
------------------------
 t
(1 row)

SELECT has_function_privilege('regress_log_fdw_nsuperuser',
    'list_postgres_log_files()', 'EXECUTE'); -- yes
 has_function_privilege 
------------------------
 t
(1 row)

-- Ensure that something gets logged to the server log file
SELECT 1/0;
ERROR:  division by zero
-- Ensure there exists at least one server log file to work on
SElECT pg_rotate_logfile();
 pg_rotate_logfile 
-------------------
 t
(1 row)

DO $$
DECLARE
    log_file INT := 0;
    loops INT := 0;
BEGIN
    LOOP
        log_file := COUNT(*) FROM list_postgres_log_files();
        -- -- Check if both .csv and .log files get created
        -- Check if .log file get created
        IF log_file >= 1 OR loops > 180 * 100 THEN EXIT; END IF;
        PERFORM pg_sleep(0.01);
        loops := loops + 1;
    END LOOP;
END
$$;
-- Check if .csv file gets created
-- SELECT COUNT(*) > 0 AS OK FROM list_postgres_log_files()
--     WHERE file_name LIKE '%.csv';
-- Check if .log file gets created
SELECT COUNT(*) > 0 AS OK FROM list_postgres_log_files()
    WHERE file_name LIKE '%.log';
 ok 
----
 t
(1 row)

-- Get .csv server log file name and create a foreign table for it
-- DO $$
-- DECLARE
--     log_file_name TEXT;
-- BEGIN
--     log_file_name := file_name FROM list_postgres_log_files()
--              WHERE file_name LIKE '%.csv' ORDER BY file_name ASC LIMIT 1;
--     PERFORM create_foreign_table_for_log_file('log_fdw_ftbl_csv','log_fdw_server', log_file_name);
-- END
-- $$;
-- Ensure that the foreign table gets created
-- SELECT COUNT(*) > 0 AS OK FROM pg_foreign_table WHERE ftrelid = 'log_fdw_ftbl_csv'::regclass;
-- Ensure that the foreign table has read data from .csv server log file
-- SELECT COUNT(*) > 0 AS OK FROM log_fdw_ftbl_csv;
-- Get .log server log file name and create a foreign table for it
DO $$
DECLARE
    log_file_name TEXT;
BEGIN
    log_file_name := file_name FROM list_postgres_log_files()
             WHERE file_name LIKE '%.log' ORDER BY file_name ASC LIMIT 1;
    PERFORM create_foreign_table_for_log_file('log_fdw_ftbl_log','log_fdw_server', log_file_name);
END
$$;
-- Ensure that the foreign table gets created
SELECT COUNT(*) > 0 AS OK FROM pg_foreign_table WHERE ftrelid = 'log_fdw_ftbl_log'::regclass;
 ok 
----
 t
(1 row)

-- Ensure that the foreign table has read data from .log server log file
SELECT COUNT(*) > 0 AS OK FROM log_fdw_ftbl_log;
 ok 
----
 t
(1 row)

-- Clean up
DROP EXTENSION log_fdw CASCADE;
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to server log_fdw_server
drop cascades to foreign table log_fdw_ftbl_log
